Kernel/Tasks/Thread.cpp

```diff
@@ -229,17 +221,10 @@ void Thread::block(Kernel::Mutex& lock, SpinlockLocker<Spinlock<LockRank::None>>
     SpinlockLocker scheduler_lock(g_scheduler_lock);
     SpinlockLocker block_lock(m_block_lock);
 
-    switch (state()) {
-    case Thread::State::Stopped:
-        // It's possible that we were requested to be stopped!
```

Comment:
I wonder why we had a special case for `Stopped` here. Are you sure this can't happen somehow?

Reply:
So I think this should be safe as long as all threads stop themselves (which they do)
and immediately yield without calling `block{,_impl}` anywhere.
Most places seem to yield after `set_state(Stopped)`. There is one exception though: https://github.com/SerenityOS/serenity/blob/3735606ddc2cdb8e2d3783ba74aacfe6a3a469dd/Kernel/Tasks/Scheduler.cpp#L306
`dispatch_one_pending_signal()` returns a `DispatchSignalResult`, which can be `Yield` if the thread should yield. I think ignoring this result is an oversight. We should maybe even make that enum `[[nodiscard]]`.

I think this call to `dispatch_one_pending_signal` at the end of `context_switch` could actually cause a SIGSTOP to be delivered and ignored. The code after that context switch could then theoretically block while `Stopped`.